### PR TITLE
add FilterFeatures input transform and tests

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR adds the possibility for using GPs in a `ModelListGP` sharing only subsets of features. This is useful in multiobjective optimization if it is known that a feature is not important for one objective but has to be taken into account for another objective.

It is implemented by adding a new input transform called `FilterFeatures`.

Addresses https://github.com/pytorch/botorch/issues/1028

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I added tests for the new transform according to the tests implemented for the other input transforms.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/pytorch/botorch, and link to your PR here.)

### CODE REVIEW

File: botorch/models/transforms/input.py [hunk 1]
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+                transform when called from within a `fantasize` call. Default: True.
+        """
+        super().__init__()
+        if feature_indices.dim() != 1:
+            raise ValueError("`feature_indices` must be a onedimensional tensor!")
+        if feature_indices.dtype != torch.int64:
+            raise ValueError("`feature_indices` tensor must be int64/long!")
+        if (feature_indices < 0).any():
+            raise ValueError(
+                "Elements of `feature_indices` have to larger/equal to zero!"
```
Comment (Balandat) [inline botorch/models/transforms/input.py:875]:
```suggestion
                "Elements of `feature_indices` have to be larger/equal to zero!"
```

File: botorch/models/transforms/input.py [hunk 2]
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+            transform_on_train: A boolean indicating whether to apply the
+                transforms in train() mode. Default: True.
+            transform_on_eval: A boolean indicating whether to apply the
+                transform in eval() mode. Default: True.
+            transform_on_fantasize: A boolean indicating whether to apply the
+                transform when called from within a `fantasize` call. Default: True.
+        """
+        super().__init__()
+        if feature_indices.dim() != 1:
+            raise ValueError("`feature_indices` must be a onedimensional tensor!")
```
Comment (Balandat) [inline botorch/models/transforms/input.py:870]:
```suggestion
            raise ValueError("`feature_indices` must be a one-dimensional tensor!")
```

File: botorch/models/transforms/input.py [hunk 3]
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+    def equals(self, other: InputTransform) -> bool:
+        r"""Check if another input transform is equivalent.
+
+        Args:
+            other: Another input transform
+
+        Returns:
+            A boolean indicating if the other transform is equivalent.
+        """
+        return (
```
Comment (Balandat) [inline botorch/models/transforms/input.py:905]: Could you add a test for this as well, please?

File: botorch/models/transforms/input.py [hunk 4]
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+
+    def transform(self, X: Tensor) -> Tensor:
+        r"""Transform the inputs by keeping only the in `feature_indices` specified
+        feature indices and filtering out the others.
+
+        Args:
+            X: A `batch_shape x q x d`-dim tensor of inputs.
+
+        Returns:
+            A `batch_shape x q x e`-dim tensor of filtered inputs, where is the length of `feature_indices`.
```
Comment (Balandat) [inline botorch/models/transforms/input.py:892]: this line will probably be too long for flake?

**Review by Balandat — COMMENTED** [review]
Thanks for putting this up, looks great. There is one little test to add, and then code formatting is failing, likely b/c this is based on some old base commit. The easiest way to fix this is probably to merge `main` back in there and make make sure to use the code formatter versions as they will be pinned in #1046.

File: botorch/models/transforms/input.py (hunk 3 above)
Comment (jduerholt) [reply botorch/models/transforms/input.py]: I added the test and fixed a bug in the equals method.

File: botorch/models/transforms/input.py (hunk 4 above)
Comment (jduerholt) [reply botorch/models/transforms/input.py]: Fixed also this.

**Comment by jduerholt** [comment]
Hi @Balandat,

thanks for reviewing. I addressed your comments. I am not so fluent with all these formatters. I hope that it passes now ;)

Best,

Johannes

**Comment by Balandat** [comment]
All good, thank you! Will get this landed soon.

**Pull request closed without merging**